client/login/controller.js

```diff
@@ -159,3 +157,34 @@ export function redirectJetpack( context, next ) {
 	}
 	next();
 }
+
+/**
+ * A middleware that enables (or disables) server side rendering for the /log-in page.
+ *
+ * Unlike the rest of the SSRed pages, the log-in page enables SSRing also when a set of parameters is set (see below
+ * whitelistedQueryKeys). Some of these parameters may need to fulfill additional formats (example: when redirect_to is
+ * present, then it also needs to start with a certain prefix).
+ *
+ * @param {object}   context  The entire request context
+ * @param {function} next     Next middleware in the running sequence
+ */
+export function setShouldServerSideRenderLogin( context, next ) {
+	const whitelistedQueryKeys = [ 'client_id', 'signup_flow', 'redirect_to' ];
+	const queryKeys = Object.keys( context.query );
+
+	// if there are any parameters, they must be ONLY the ones in the whitelist
+	const hasOnlyValidKeys =
+		queryKeys.length === intersection( queryKeys, whitelistedQueryKeys ).length;
+
+	let hasValidRedirectTo = true;
+	if ( hasOnlyValidKeys && context.query.redirect_to ) {
+		const redirectToDecoded = decodeURIComponent( context.query.redirect_to );
+		hasValidRedirectTo =
+			redirectToDecoded.startsWith( 'https://wordpress.com/theme' ) ||
+			redirectToDecoded.startsWith( 'https://wordpress.com/go' );
+	}
```

Comment:
I understand that this list of valid `redirect_to` params is expected to grow quite large. Have you thought about how that should be handled? Maintaining a _very large list of `||` conditions_ doesn't appeal 😉 

---

I'd prefer avoiding the mutation. It also occurs to me that short-circuiting on `hasOnlyValidKeys` is nice and saves a potentially expensive set of checks, we may put `hasValidRedirectTo` into a state that's not accurate 🙂 

We might improve that if we pull `hasValidRedirectTo` out into a separate function, then we could keep the short circuiting and semantic correctness of `hasValidRedirectTo`:

```js
function isValidRedirectTo( redirectTo ) {
  if ( 'undefined' === typeof redirectTo ) {
    return true;
  }
  const decoded = decodeURIComponent( redirectTo );

  return decoded.startsWith( 'https://wordpress.com/theme' ) || …
}

export function setShouldServerSideRenderLogin( context, next ) {
  // …
  context.serverSideRender = hasOnlyValidKeys && isValidRedirectTo( context.query.redirect_to );
  next();
}
```

Reply:
Like it, will modify it this way!
----
Regarding the ||, it feels keeping it this way for 2 values is the best variant due to the readability (i.e. if it starts with this or that, English-like). Once more values come in, depending on what those values really represent, this can evolve into something more complex (or can be combined with some Regexps).